Add feature flag to enable inventory

rioug opened the pull request:
⚠️ **Please use clockify code # 76  Network 2.0**

#### What? Why?

- Closes #13373 

This some preliminary work before adding tag rules to variant tag. We only want to have tag rule enable for either variant override or variant tag. We don't want to have to support a mix of the two. So the first step is to put the inventory feature (powererd by variant override) behind a feature tag. 

~~**Inventory will be fully enabled for now.**~~
<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->


#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->
This is by no mean exhaustive, below I listed the area that I am aware of that are impacted by the inventory but I might have missed something.

##### Inventory and product import
With inventory disabled, as an Enterprise user:
- Log in the back office  
- Navigate to the Bulk Edit Product page 
 --> Inventory should not be available in the sub menu
- Navigate to the Product Import page
  --> Inventories should not be available as an import type and the inventory CVS template should not be available 
- Navigate to the enterprise settings page
  --> "Inventory settings" should not be available in the left hand side menu

With inventory enabled, as an Enterprise user:
- Log in the back office  
- Navigate to the Bulk Edit Product page 
- Click on Inventory in the sub menu
 --> Inventory page should load 
- Navigate to the Product Import page
  --> Inventories be available as an import type and the inventory CVS template should be visible 
- Import product into inventories
  --> import should succeed without any errors
- Navigate to the enterprise settings page
- Click on "Inventory settings" on the left hand side menu
  --> "Inventory settings" page should load

##### Shop front
With inventory enabled, as an Enterprise user :
- create a variant override, with a different price (update price in the inventory for a given variant)

With inventory enabled, as a customer : 
- Navigate to the shop with the variant override 
  --> it should display the variant override price
- Place an order with said override
  --> you should be able to place an order without any error

With inventory disabled, as a customer : 
- Navigate to the shop with the variant override 
  --> it should display the variant price, ie it should not be overridden
- Place an order with the variant
  --> you should be able to place an order without any error

##### Report
With inventory enabled, as an Enterprise user
 - Navigate to Reports page 
 - Choose "Orders & Fulfillment Reports" -> "Order Cycle Customer Totals" 
 - Generate a report including the variant that was overridden
   --> The report should include at least 2 lines from the 2 orders with placed above. One should have the overridden price and one should have the normal price   
  
With inventory disabled, as an Enterprise user, repeat the scenario above. You should get the same result, as the inventory settings should not affect existing orders.

##### Tag rules with variant 
With inventory disabled, as an Enterprise user
- Navigate to enterprise settings page and click on "tag rules" in the left hand side menu
- Click on  "+ Add New Default Rule"
  --> the rule type dropdown should not include "Show or Hide variant in my shopfront"

With inventory enabled, as an Enterprise user
- Navigate to enterprise settings page and click on "tag rules" in the left hand side menu
- Click on  "+ Add New Default Rule"
  --> the rule type dropdown should include "Show or Hide variant in my shopfront"
- Select "Show or Hide variant in my shopfront" and Click "Add Rule"
- Add a tag 
- Save by clicking on "Update" at the bottom of the page.
- Navigate to the Bulk Edit product page, and then to the Inventory
- Add the same tag on one of the variant
We just added a rule that hides variant which have a variant override with a tag matching the one define in the rule.

With inventory enabled, as a Customer 
- Load the shop front containing the variant we just added a tag to.
  -- > the variant shouldn't be displayed 

With inventory disabled. as a Customer
- Load the shop front containing the variant we just added a tag to.
  -- > the variant should be displayed 

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [x] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.

mkllnk commented on app/controllers/admin/subscription_line_items_controller.rb:
```diff
@@ -53,7 +53,11 @@ def price_estimate
 
       fee_calculator = OpenFoodNetwork::EnterpriseFeeCalculator.new(@shop, @order_cycle)
-      OpenFoodNetwork::ScopeVariantToHub.new(@shop).scope(@variant)
+
+      if OpenFoodNetwork::FeatureToggle.enabled?(:inventory, @shop)
+        OpenFoodNetwork::ScopeVariantToHub.new(@shop).scope(@variant)
+      end
+
```
Seeing a lot of these changes, I was wondering if it would be easier to just change the `scope` method? We use that everywhere, don't we?

mkllnk reviewed (commented):
Wow, this is a big piece of work and it looks solid.

I have to admit that I'm a bit nervous about this path though. Deactivating the inventory completely may show a lot of bugs or frustration when people suddenly sell products at different prices. But I guess that there will be more work communicating the transition or even copying data into the new structure. This is just a first technical step.

There are a couple of things that I would change here though.

* Invert the feature toggle. It would result in an ugly negative feature toggle. You enable a toggle to disable a feature. But my reason is more pragmatic: Flipper is built for development in one direction. It can't disable features for individual users.
* Move a lot of conditionals into the scoper module. It knows the distributor and can set a simple boolean in the constructor. It will result in a smaller diff and it's easier to change when we touch the feature toggle.

Apart from these minor tweaks, this looks really good.

dacook commented on app/controllers/admin/product_import_controller.rb:
```diff
@@ -36,7 +39,22 @@ def validate_data
+        },
+        updated_ids: @importer.updated_ids,
+        errors: @importer.errors.full_messages
+      }
```
Nice, it's probably better to have the json format defined here in the controller than in the class 👍

dacook commented on app/controllers/admin/enterprises_controller.rb:
```diff
@@ -375,6 +378,19 @@ def load_properties
+      @tag_rule_types = [
+        { id: "FilterShippingMethods", name: t('js.tag_rules.show_hide_shipping') },
+        { id: "FilterPaymentMethods", name: t('js.tag_rules.show_hide_payment') },
+        { id: "FilterOrderCycles", name: t('js.tag_rules.show_hide_order_cycles') }
```
Now these translations might not need to be under the `js` namespace. 

Question, do we want these tag rule types to be defined in ruby in the long run, or would we move them back to JS later? 
I'm not really sure, but given that they are directly related to the angular implementation, I think they could move back to JS later. Did you have anything else in mind?

dacook reviewed (changes requested):
Great job.

I agree with Maikel's comment:
> Move a lot of conditionals into the scoper module. It knows the distributor and can set a simple boolean in the constructor. It will result in a smaller diff and it's easier to change when we touch the feature toggle.

This seems safer, and worth changing before doing a big round of manual testing.

dacook commented:
Regarding rollout: it seems pointless merging this if we're not able to start progressively disabling it.

 > Flipper is built for development in one direction. It can't disable features for individual users.

Therefore if it's enabled by default, we can't disable it for certain users, we can only disable the entire instance for testing? 

I'm guessing that the first step should be to disable for new users. In that case, I'd suggest we keep the current condition (enabled means enabled), and roll out like this:

1. Disabled by default (for new users); with migration to enable for all current users.
2. Use a script or migration to disable for all current users that aren't using it
3. Customer support work with each remaining user to disable when ready

(steps 1+2 could be combined, but it's probably safer to separate them out)

mkllnk commented:
> Disabled by default (for new users); with migration to enable for all current users.

I think that we can keep it enabled to start with. It would be good if instances could decide when to start the migration.

Maybe, instead of a migration, we should write a rake task that can enable a feature for all current users. After running that, we can disable the feature and safely disable individually.

In this case, we are enabling on enterprise level though. So maybe we need to be able to define the class on the rake task, too.

```
rake ofn:feature:enable[inventory,Enterprise]
```

> Use a script or migration to disable for all current users that aren't using it.

This would need to be a special migration or rake task because it's very specific to this feature. It might be handy to reduce the noise and focus on the enterprises actually using inventory. If it's a script then it can be repeated when enterprises stop using inventory. It would be the safest way to disable the feature for anyone. Otherwise you suddenly change the prices in the shop front when switching inventory off.

```
rake ofn:feature:prune_inventory_use
```

RachL commented:
> It would be good if instances could decide when to start the migration.

I wouldn't leave that decision to instance managers. This would blow delays. Here is my proposal :

1. Disabled by default (for new users); with migration to enable for all current users.
2. List all current users who are not using it and send this list to instance managers. Tell them they have 2 weeks to confirm with these users that they are indeed not using inventory
3. Use a script or migration to disable for all current users that aren't using it
3. Release tags with tags rules on product page.
4. Customer support work with each remaining user to disable when ready

rioug commented:
I updated the PR to disable `inventory` feature by default, I also added a group so we can enable `inventory` for all enterprises created before 2025-7-4, I am assuming recent enterprise are not using inventory. The date can be changed if needed.
 
>     1. Disabled by default (for new users); with migration to enable for all current users.

We'll need to enable inventory for  `enterprise_created_before_2025_07_04 group`. This can be done with Flipper CLI : 
```
bundle exec flipper enable -g enterprise_created_before_2025_07_04 :inventory
```

>     2. List all current users who are not using it and send this list to instance managers. Tell them they have 2 weeks to confirm with these users that they are indeed not using inventory
 
>     3. Use a script or migration to disable for all current users that aren't using it
We can again use the flipper CLI here 
- disable inventory for `enterprise_created_before_2025_07_04` group : `bundle exec flipper disable -g enterprise_created_before_2025_07_04 :inventory`
- enabled it for enterprise that are still using it :
```
 bundle exec flipper enable -a Enterprise;10,Enterprise;20,Enterprise:30 :inventory
```
We can only add up to 100 enterprises this way see : https://www.flippercloud.io/docs/features/actors#limitations. FYI on `fr_prod` there are 368 enterprises using inventory. We could increase the number if needed but it can create performance issue.  

>     4. Release tags with tags rules on product page.
 
>     5. Customer support work with each remaining user to disable when ready
It would just be a matter of manually removing the enterprise in flipper when ready to disable.

rioug commented on lib/reporting/reports/orders_and_fulfillment/base.rb:
```diff
@@ -80,10 +80,16 @@ def total_units(line_items)
-              report_variant_overrides[distributor_id] || {},
+              variant_overrides,
             )
         end
```
I am pretty sure this is covered somewhere else, possibly by system spec, but I'll have to double check.

rioug commented on app/controllers/admin/subscription_line_items_controller.rb:
```diff
@@ -54,9 +54,9 @@ def price_estimate
-        OpenFoodNetwork::ScopeVariantToHub.new(@shop).scope(@variant)
-      end
+      OpenFoodNetwork::ScopeVariantToHub.new(@shop).scope(
+        @variant, inventory_enabled: OpenFoodNetwork::FeatureToggle.enabled?(:inventory, @shop)
```
I totally misunderstood his comment, in retrospect it makes a lot of sense. Fixed in this commit : https://github.com/openfoodfoundation/openfoodnetwork/pull/13381/commits/b7f969eed90f47e34500d6f0ed31fe45c4293938#diff-32a6ee84a122e56dac0e9373626b905319cc5d54342891a28aa9c003ba6566bc
The change actually uncovered some scenarios I had missed.

dacook replied:
Yay!

rioug commented:
> Also for step 1, we should be able to do that automatically as part of this PR (see for example [this migration](https://github.com/openfoodfoundation/openfoodnetwork/blob/aeac3e500bbe654769e151b9c7daabf064326020/db/migrate/20231129222807_enable_feature_backround_reports.rb#L5))

Done, migration added. I am not sure why I didn't think of that.

dacook reviewed (approved):
Yay, looks good to me!

rioug commented on lib/reporting/reports/orders_and_fulfillment/base.rb:
```diff
@@ -81,16 +81,14 @@ def total_units(line_items)
-              distributor_id,
-              variant_overrides,
-            )
+            OpenFoodNetwork::ScopeVariantToHub.new(distributor, variant_overrides)
```
I did that because of this change: https://github.com/openfoodfoundation/openfoodnetwork/pull/13381/commits/b7f969eed90f47e34500d6f0ed31fe45c4293938#diff-32a6ee84a122e56dac0e9373626b905319cc5d54342891a28aa9c003ba6566bc
 
Passing the id breaks the call `OpenFoodNetwork::FeatureToggle.enabled?` , I though it was better to be consistent and pass the object instead to handling the unusual scenario in `ScopeVariantToHub`

rioug commented on db/migrate/20250709012346_enable_feature_inventory_for_existing_enterprises.rb:
```diff
@@ -0,0 +1,13 @@
+class EnableFeatureInventoryForExistingEnterprises < ActiveRecord::Migration[7.0]
+  # rubocop:disable Naming/VariableNumber
+  def up
+    Flipper.enable_group(:inventory, :enterprise_created_before_2025_07_04)
```
That was also my initial idea, but doing so can potentially create some performance issue see my earlier comment :

> We can only add up to 100 enterprises this way see : https://www.flippercloud.io/docs/features/actors#limitations. FYI on fr_prod there are 368 enterprises using inventory. We could increase the number if needed but it can create performance issue.

rioug commented on app/views/admin/order_cycles/_advanced_settings.html.haml:
```diff
@@ -7,9 +7,10 @@
-    .four.columns
-      = f.radio_button :preferred_product_selection_from_coordinator_inventory_only, true
-      = f.label :preferred_product_selection_from_coordinator_inventory_only, t('.preferred_product_selection_from_coordinator_inventory_only_here')
+    - if feature?(:inventory, @order_cycle.coordinator)
```
make sense, I just did that here : https://github.com/openfoodfoundation/openfoodnetwork/pull/13381/commits/6ba80f57b3f7f8851110b4db2c9b3f1fbfb5bcd8

filipefurtad0 commented:
Hey @rioug,

Just to confirm, this PR disables the inventory for enterprises created after 2025-07-04. So, if an enterprise is created at the 07-05 and sets up inventory, then when we deploy this PR in production the feature will be disabled, right?

If so, this could happen not only to new enterprises (created after this PR is in prod), but also for enterprises created between 2025-07-04 and the deployment date (sometime next week), right?
